File: server/models/usermodel.js
const mongoose = require("mongoose");
const validator = require("validator");

const userSchema = new mongoose.Schema({
  name: {
    type: String,
    required: [true, "Name is a required field"],
  },
  email: {
    type: String,
    required: [true, "Please provide your email"],
    unique: true,
    lowercase: true,
    validate: [validator.isEmail, "Please check the email and try again"],
  },
  username: {
    type: String,
    required: [true, "Username is required"],
    unique: true,
  },
  password: {
    type: String,
    required: [true, "Password is required"],
    minlength: 8,
    select: false,
  },
  currentHostel: {
    type: String,
    default: undefined,
  },
  photo: {
    type: String,
    required: [true, "Photo of user is a required field"],
  },
  phoneNumber: {
    type: String,
    required: [true, "Phone number is required"],
    validate: {
      validator: function (v) {
        return /\d{10,15}/.test(v);
      },
      message: (props) => `${props.value} is not a valid phone number!`,
    },
  },
  referenceNumber: {
    type: String,
    required: [true, "Reference number is a required field"],
  },
  motherName: {
    type: String,
    default: undefined,
  },
  fatherName: {
    type: String,
    default: undefined,
  },
  guideanContact: {
    type: String,
    default: undefined,
  },
  gender: {
    type: String,
    enum: ["Male", "Female"],
    required: true,
  },
  bYear: {
    type: Number,
    required: [true, "Please provide birth year"],
  },
  bDay: {
    type: Number,
    required: [true, "Please provide day"],
  },
  bMonth: {
    type: Number,
    required: [true, "Please provide month"],
  },
  birthCert: {
    type: String,
    required: [true, "A scanned copy of birth certificate is required"],
  },
  homeTown: {
    type: String,
    default: undefined,
  },
  criminalRecords: {
    type: String,
    default: "none",
  },
  shsAttended: {
    type: String,
    required: true,
  },
  role: {
    type: String,
    enum: ["student", "admin", "driver"],
    default: "student",
  },
  otherRecords: {
    type: String,
    default: undefined,
  },
  descriptor: {
    type: Buffer,
    required: true,
  },
});

const User = mongoose.model("User", userSchema);
module.exports = User;

File: lib/utils/models/user_model.dart
import 'dart:convert';

// ignore_for_file: public_member_api_docs, sort_constructors_first
class UserModel {
  final String username;
  final String password;
  final String email;
  final String photo;
  final int bDay;
  final String currentHostel;
  final int bYear;
  final int bMonth;
  final String gender;
  final String parentPhone;
  final String birthCert;
  final String name;
  final String phoneNumber;
  final String referenceNumber;
  final String fatherName;
  final String motherName;
  final String hometown;
  final String crimeRecords;
  final String role;
  final String otherRecords;
  final String shsAttended;

  UserModel({
    required this.username,
    required this.password,
    required this.email,
    required this.photo,
    required this.bDay,
    required this.currentHostel,
    required this.bYear,
    required this.bMonth,
    required this.gender,
    required this.parentPhone,
    required this.birthCert,
    required this.name,
    required this.phoneNumber,
    required this.referenceNumber,
    required this.fatherName,
    required this.motherName,
    required this.hometown,
    required this.crimeRecords,
    required this.role,
    required this.otherRecords,
    required this.shsAttended,
  });

  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      'username': username,
      'password': password,
      'email': email,
      'photo': photo,
      'bDay': bDay,
      'currentHostel': currentHostel,
      'bYear': bYear,
      'bMonth': bMonth,
      'gender': gender,
      'parentPhone': parentPhone,
      'birthCert': birthCert,
      'name': name,
      'phoneNumber': phoneNumber,
      'referenceNumber': referenceNumber,
      'fatherName': fatherName,
      'motherName': motherName,
      'hometown': hometown,
      'crimeRecords': crimeRecords,
      'role': role,
      'otherRecords': otherRecords,
      'shsAttended': shsAttended,
    };
  }

  factory UserModel.fromMap(Map<String, dynamic> map) {
    return UserModel(
      username: map['username'] as String,
      password: map['password'] as String,
      email: map['email'] as String,
      photo: map['photo'] as String,
      bDay: map['bDay'] as int,
      currentHostel: map['currentHostel'] as String,
      bYear: map['bYear'] as int,
      bMonth: map['bMonth'] as int,
      gender: map['gender'] as String,
      parentPhone: map['parentPhone'] as String,
      birthCert: map['birthCert'] as String,
      name: map['name'] as String,
      phoneNumber: map['phoneNumber'] as String,
      referenceNumber: map['referenceNumber'] as String,
      fatherName: map['fatherName'] as String,
      motherName: map['motherName'] as String,
      hometown: map['hometown'] as String,
      crimeRecords: map['crimeRecords'] as String,
      role: map['role'] as String,
      otherRecords: map['otherRecords'] as String,
      shsAttended: map['shsAttended'] as String,
    );
  }

  String toJson() => json.encode(toMap());

  factory UserModel.fromJson(String source) =>
      UserModel.fromMap(json.decode(source) as Map<String, dynamic>);
}
